core/os/mutex.h

```diff
@@ -66,6 +67,11 @@ class MutexImpl {
 	_ALWAYS_INLINE_ bool try_lock() const {
 		return mutex.try_lock();
 	}
+#else
+	_ALWAYS_INLINE_ void lock() const {}
+	_ALWAYS_INLINE_ void unlock() const {}
+	_ALWAYS_INLINE_ bool try_lock() const { return true; }
+#endif
```

Comment:
I know you went for the most surgical changes to get things to compile first, which is a great approach.

To polish further now that it works, I think we should make intent clear in the API by more explicitly bypassing threading APIs. What I mean is typically (at least for short classes/functions) is to fully enclose the class/function declaration in the macros, and provide a fallback/dummy version for the no-threads branch.

In this case, I would take it as far as just having two branches for everything declared in that file.

```cpp
#ifdef THREADS_ENABLED

#ifdef MINGW_ENABLED
#define MINGW_STDTHREAD_REDUNDANCY_WARNING
#include "thirdparty/mingw-std-threads/mingw.mutex.h"
#define THREADING_NAMESPACE mingw_stdthread
#else
#include <mutex>
#define THREADING_NAMESPACE std
#endif

template <class MutexT>
class MutexLock;

template <class StdMutexT>
class MutexImpl {
	friend class MutexLock<MutexImpl<StdMutexT>>;
	...
};

...

using Mutex = MutexImpl<THREADING_NAMESPACE::recursive_mutex>; // Recursive, for general use
using BinaryMutex = MutexImpl<THREADING_NAMESPACE::mutex>; // Non-recursive, handle with care
extern template class MutexImpl<THREADING_NAMESPACE::recursive_mutex>;
extern template class MutexImpl<THREADING_NAMESPACE::mutex>;
extern template class MutexLock<MutexImpl<THREADING_NAMESPACE::recursive_mutex>>;
extern template class MutexLock<MutexImpl<THREADING_NAMESPACE::mutex>>;

#else // No threads.

template <class MutexT>

class MutexLock;
template <class StdMutexT>

class MutexImpl {
	// stubs
	...
};

...

#endif // THREADS_ENABLED
```

There might be some repetition in the overall structure, but I think it's clearer for the reader to see all these as completely separate implementations of the same interface, instead of Frankeinstein classes where only parts are compile flag specific.

Discussion welcome, CC @Faless

Reply:
I replaced the Frankeinstein cutout with two different parts.